Documentation/design-docs/native-hosting-initialize.md

```diff
@@ -0,0 +1,213 @@
+* `dotnet_root` - path to the root of the .NET Core installation in use. This typically points to the install location from which the `hostfxr` has been loaded. For example on Windows this would typically point to `C:\Program Files\dotnet`. The path is used to search for shared frameworks and potentially SDKs.
+
+
+``` C++
+int hostfxr_initialize_for_app(
+    int argc,
+    const char_t * argv[],
+    const char_t * app_path  ,
+    const hostfxr_initialize_parameters * parameters,
+    hostfxr_handle * host_context_handle
+);
+```
+
+Initializes the hosting components for running a managed application.
+When used to execute an app, the `app_path` (or CLI equivalent) will be used to locate the `.runtimeconfig.json` and the `.deps.json` which will be used to load the application and its dependent frameworks.
+* `argc` and `argv` - the command line - entirely optional
+* `app_path` - path to the application (the managed `.dll`) to run. This can be `nullptr` if the app is specified in the command line arguments.
+* `parameters` - additional parameters - see `hostfxr_initialize_parameters` for details. (Could be made optional potentially)
+* `host_context_handle` - output parameter. On success receives an opaque value which identifies the initialized host context. The handle should be closed by calling `hostfxr_shutdown`.
+
+This function can only be called once per-process. It's not supported to run multiple apps in one process (even sequentially).
+
+This function will fail if there already is a CoreCLR running in the process.
+
+*Note: This is effectively a replacement for `hostfxr_main_startupinfo` and `hostfxr_main`. Currently it is not be a goal to fully replace these APIs because they also support SDK commands which are special in lot of ways and don't fit well with the rest of the native hosting. There's no scenario right now which would require the ability to issue SDK commands from a native host. That said nothing in this proposal should block enabling even SDK commands through these APIs.*
+
+
+``` C++
+int hostfxr_initialize_for_runtime_config(
+    const char_t * runtime_config_path,
+    const hostfxr_initialize_parameters * parameters,
+    hostfxr_handle * host_context_handle
+);
+```
+
+This function would load the specified `.runtimeconfig.json`, resolve all frameworks, resolve all the assets from those frameworks abd then prepare runtime initialization where the TPA contains only frameworks. Note that this case does NOT consume any `.deps.json` from the app/component (only processes the framework's `.deps.json`). This entry point is intended for `comhost`/`ijwhost`/`nethost` and similar scenarios.
```

Comment:
Instead of consuming a path, I would suggest we consume the read in configuration file. This would allow embedding the configuration as well as allow us to avoid assumptions about how to access the environment's file system. Is there a reason we want to always read from disk? Might be interesting for ASP.NET or other scenarios to pipe in config from somewhere and with this API they would have to write to disk. Instead they could just pass a buffer with the configuration contents?

Reply:
Depending on discussions around other APIs, this might be the main way for component hosting (the easy to use API seems to be hard to specify... we may not be able to get one in time). In such case having to load the file into memory seems like rather unnecessary burden.

The thinking was that we could add another version which loads it from memory. There's also some desire to be able to specify it as a structured parameter - not JSON, some kind of struct.